integration/spark/app/src/main/java/io/openlineage/spark/agent/OpenLineageSparkListener.java

```diff
@@ -264,30 +245,99 @@ public static void close() {
               + "Lineage events will not be collected");
+      return;
+    }
+    initializeContextFactoryIfNotInitialized(sparkEnv.conf(), appName);
+  }
+
+  private void initializeContextFactoryIfNotInitialized(SparkConf sparkConf, String appName) {
+    if (contextFactory != null || isDisabled) {
+      return;
+    }
+    try {
+      ArgumentParser args = ArgumentParser.parse(sparkConf);
+      contextFactory = new ContextFactory(new EventEmitter(args, appName));
+    } catch (URISyntaxException e) {
+      log.error("Unable to parse open lineage endpoint. Lineage events will not be collected", e);
     }
   }
 
+  public void emitApplicationEvent(Long time, OpenLineage.RunEvent.EventType eventType) {
+    OpenLineage openLineage = new OpenLineage(Versions.OPEN_LINEAGE_PRODUCER_URI);
+    EventEmitter emitter = contextFactory.openLineageEventEmitter;
+    OpenLineage.RunBuilder runBuilder =
+        openLineage.newRunBuilder().runId(emitter.getApplicationRunId());
+    if (emitter.getParentRunId().isPresent()
+        && emitter.getParentJobName().isPresent()
+        && emitter.getParentJobNamespace().isPresent()) {
+      runBuilder.facets(
+          openLineage
+              .newRunFacetsBuilder()
+              .parent(
+                  openLineage.newParentRunFacet(
+                      openLineage.newParentRunFacetRun(emitter.getParentRunId().get()),
+                      openLineage.newParentRunFacetJob(
+                          emitter.getParentJobNamespace().get(), emitter.getParentJobName().get())))
+              .build());
+    }
+    OpenLineage.RunEvent applicationEvent =
+        openLineage
+            .newRunEventBuilder()
+            .eventType(eventType)
+            .eventTime(toZonedTime(time))
+            .job(
+                openLineage
+                    .newJobBuilder()
+                    .namespace(emitter.getJobNamespace())
+                    .name(
+                        emitter
+                            .getApplicationJobName()
+                            .replaceAll(CAMEL_TO_SNAKE_CASE, "_$1")
+                            .toLowerCase(Locale.ROOT))
+                    .build())
+            .run(runBuilder.build())
+            .inputs(Collections.emptyList())
+            .outputs(Collections.emptyList())
+            .build();
+    emitter.emit(applicationEvent);
+  }
+
+  public void emitApplicationStartEvent(Long time) {
+    emitApplicationEvent(time, OpenLineage.RunEvent.EventType.START);
+  }
+
+  public void emitApplicationEndEvent(Long time) {
+    emitApplicationEvent(time, OpenLineage.RunEvent.EventType.COMPLETE);
+  }
+
   private static boolean checkIfDisabled() {
     String isDisabled = Environment.getEnvironmentVariable("OPENLINEAGE_DISABLED");
     return Boolean.parseBoolean(isDisabled);
   }
+
+  protected ZonedDateTime toZonedTime(long time) {
+    Instant i = Instant.ofEpochMilli(time);
+    return ZonedDateTime.ofInstant(i, ZoneOffset.UTC);
```

Comment:
shouldn't we use system time zone or sparkContext timezone if it is set?

Reply:
I've just copied existing method from execution contexts. Now removed it and added `TimeUtils`.